internal/license/agent.go

```diff
@@ -2,126 +2,204 @@ package license
+	defer ticker.Stop()
+
 	for {
 		select {
+		case <-ticker.C:
+			a.logger.V(util.DebugLevel).Info("retrieving license from external service")
+			if err := a.reconcileLicenseWithKonnect(ctx); err != nil {
+				a.logger.Error(err, "could not reconcile license with Konnect")
+			}
+			// Reset the ticker to run with the expected period which may change after we receive the license.
+			ticker.Reset(a.resolvePollingPeriod())
 		case <-ctx.Done():
 			a.logger.Info("context done, shutting down license agent")
-			a.ticker.Stop()
 			return ctx.Err()
-		case <-a.ticker.C:
-			a.logger.V(util.DebugLevel).Info("retrieving license from external service")
-			if err := a.updateLicense(ctx); err != nil {
-				a.logger.Error(err, "could not update license")
-			}
 		}
 	}
 }
 
-// updateLicense retrievs a license from an outside system. If it successfully retrieves a license, it updates the
-// in-memory license cache.
-func (a *Agent) updateLicense(ctx context.Context) error {
-	ctx, cancel := context.WithTimeout(ctx, PollingTimeout)
-	defer cancel()
+func (a *Agent) resolvePollingPeriod() time.Duration {
+	// If we already have a license, start with the regular polling period (happy path) ...
+	if a.cachedLicense.IsPresent() {
+		return a.regularPollingPeriod
+	}
+	// ... otherwise, start with the initial polling period which is shorter by default (to get a license faster
+	// when it appears, e.g. when a user upgrades from Free to Enterprise tier).
+	return a.initialPollingPeriod
+}
 
-	// This is an array because it's a Kong entity collection, even though we only expect to have exactly one license.
-	licenses, err := a.konnectAPIClient.List(ctx, 0)
-	if err != nil {
-		return fmt.Errorf("could not retrieve license: %w", err)
+// reconcileLicenseWithKonnect retrieves a license from upstream and caches it if it is newer than the cached license.
+// When it's the first time retrieving a license, it will always cache it.
+func (a *Agent) reconcileLicenseWithKonnect(ctx context.Context) error {
+	updatedAtAsString := func(updatedAt uint64) string {
+		return time.Unix(int64(updatedAt), 0).String()
 	}
-	if len(licenses.Items) == 0 {
-		return fmt.Errorf("received empty license response")
+
+	retrievedLicense, err := a.retrieveLicenseFromUpstream(ctx)
+	if err != nil {
+		// If the license is not found, we do not return an error as it's expected when a customer is on the Free tier.
+		if errors.Is(err, license.ErrNotFound) {
+			a.logger.V(util.DebugLevel).Info("no license found in Konnect")
+			return nil
+		}
+		return err
 	}
-	license := licenses.Items[0]
-	if license.UpdatedAt > a.license.UpdatedAt {
-		a.logger.V(util.InfoLevel).Info("updating license cache",
-			"old_updated_at", time.Unix(int64(a.license.UpdatedAt), 0).String(),
-			"new_updated_at", time.Unix(int64(license.UpdatedAt), 0).String(),
+
+	if a.cachedLicense.IsAbsent() {
+		a.logger.V(util.InfoLevel).Info("caching license retrieved from the upstream",
```

Comment:
Clarifying that this is the first population, which otherwise isn't clear unless you know the subsequent entries will always mention it's being replaced.
```suggestion
		a.logger.V(util.InfoLevel).Info("caching initial license retrieved from the upstream",
```